river/bandit/evaluate.py

```diff
@@ -0,0 +1,153 @@
+
+    >>> trace = bandit.evaluate(
+    ...     policies=[
+    ...         bandit.UCB(delta=1),
+    ...         bandit.EpsilonGreedy(epsilon=0.1, seed=42),
+    ...     ],
+    ...     env=gym.make(
+    ...         'river_bandits/CandyCaneContest-v0',
+    ...         max_episode_steps=100
+    ...     ),
+    ...     pull_func=pull_func,
+    ...     n_episodes=5,
+    ...     seed=42
+    ... )
+
+    >>> for step in trace:
+    ...     print(step)
+    ...     break
+    {'episode': 0, 'step': 0, 'policy_idx': 0, 'action': 0, 'reward': 0.0, 'reward_stat': 0.0}
+
+    The return type of this function is a generator. Each step of the generator is a dictionary.
+    You can pass the generator to a `pandas.DataFrame` to get a nice representation of the results.
+
+    >>> import pandas as pd
+
+    >>> trace = bandit.evaluate(
+    ...     policies=[
+    ...         bandit.UCB(delta=1),
+    ...         bandit.EpsilonGreedy(epsilon=0.1, seed=42),
+    ...     ],
+    ...     env=gym.make(
+    ...         'river_bandits/CandyCaneContest-v0',
+    ...         max_episode_steps=100
+    ...     ),
+    ...     pull_func=pull_func,
+    ...     n_episodes=5,
+    ...     seed=42
+    ... )
+
+    >>> trace_df = pd.DataFrame(trace)
+    >>> trace_df.sample(5, random_state=42)
+         episode  step  policy_idx  action  reward  reward_stat
+    521        2    60           1      10     0.0         30.0
+    737        3    68           1      90     1.0         43.0
+    740        3    70           0       0     0.0         14.0
+    660        3    30           0       0     0.0         10.0
+    411        2     5           1      31     0.0          2.0
+
+    The length of the dataframe is the number of policies times the number of episodes times the
+    maximum number of steps per episode.
+
+    >>> len(trace_df)
+    1000
+
+    >>> (
+    ...     trace_df.policy_idx.nunique() *
+    ...     trace_df.episode.nunique() *
+    ...     trace_df.step.nunique()
+    ... )
+    1000
+
+    """
+
+    reward_stat = reward_stat or stats.Sum()
+    rng = random.Random(seed)
+
+    for episode in range(n_episodes):
+        episode_policies = [policy.clone() for policy in policies]
+        episode_env = copy.deepcopy(env)
+        episode_env.reset(seed=rng.randint(0, 2**32))
+        episode_env.action_space.seed(rng.randint(0, 2**32 - 1))
+        episode_envs = [copy.deepcopy(episode_env) for _ in episode_policies]
+        episode_reward_stats = [reward_stat.clone() for _ in policies]
+
+        step = 0
+        done = False
+
+        while not done:
+            terminated, truncated = False, False
+            for i, (_policy, _env, _reward_stat) in enumerate(
+                zip(episode_policies, episode_envs, episode_reward_stats)
+            ):
+
+                action = pull_func(_policy, _env)
+                observation, reward, terminated, truncated, info = _env.step(action)
+                _reward_stat.update(reward)
```

Comment:
You have several times the same environment instance sequentially.

In this case, one of the environments may end before the others.
This is not handled; the finished environment is often reset, and the steps are continued.

Reply:
Ah yes good point. I've only worked with envs that end after a fixed number of episodes. But you're right, we should handle the general case.